src/Lib/App/Error.hs

```diff
@@ -1,10 +1,104 @@
+{-# LANGUAGE ConstraintKinds #-}
+
 module Lib.App.Error
        ( AppError (..)
+       , WithError
+       , IError
+       , throwOnNothingM
+       , notFoundOnNothingM
+       , isServerError
+       , isNotAllowed
+       , isInvalid
+       , notFound
+       , serverError
+       , notAllowed
+       , invalid
+       , headerDecodeError
+       , jobDecodeError
+       , toHttpError
        ) where
 
-data AppError =
-    Invalid Text
-  | NotAllowed Text
-  | NotFound
-  | ServerError Text
-  deriving (Show, Eq)
+import Control.Monad.Except (MonadError, throwError)
+import Servant.Server (ServantErr, err401, err404, err417, err500, errBody)
+
+newtype AppError = InternalError IError deriving (Show, Eq)
+
+data IError =
+    -- | General not found
+      NotFound
+    -- | Some exceptional circumstance has happened
+    -- stop execution and return. Optional text to
+    -- provide some context in server logs
+    | ServerError Text
+    -- | A required permission level was not met.
+    -- Optional text to provide some context.
+    | NotAllowed Text
+    -- | Given inputs do not conform to the expected
+    -- format or shape. Optional text to
+    -- provide some context in server logs
+    | Invalid Text
+    -- | An authentication header that was required
+    -- was provided but not in a format that the server
+    -- can understand
+    | HeaderDecodeError
+    -- | Failed to parse 'CDMP.Effects.Job.Job' from 'Text'.
+    | JobDecodeError Text
+    deriving (Show, Eq)
+
+isServerError :: AppError -> Bool
+isServerError (InternalError (ServerError _)) = True
+isServerError _                               = False
+
+isNotAllowed :: AppError -> Bool
+isNotAllowed (InternalError (NotAllowed _)) = True
+isNotAllowed _                              = False
+
+isInvalid :: AppError -> Bool
+isInvalid (InternalError (Invalid _)) = True
+isInvalid _                           = False
+
+----------------------------------------------------------------------------
+-- Internal Error helpers
+----------------------------------------------------------------------------
+
+notFound :: AppError
+notFound = InternalError NotFound
+
+serverError :: Text -> AppError
+serverError = InternalError . ServerError
+
+notAllowed :: Text -> AppError
+notAllowed = InternalError . NotAllowed
+
+invalid :: Text -> AppError
+invalid = InternalError . Invalid
+
+headerDecodeError :: AppError
+headerDecodeError = InternalError HeaderDecodeError
+
+jobDecodeError :: Text -> AppError
+jobDecodeError = InternalError . JobDecodeError
+
+-- | Type alias for errors.
+type WithError m = MonadError AppError m
```

Comment:
Could you please move this at the top of this module how it's done in CDMP?

* https://github.com/Holmusk/CDMP/blob/cd24a6fb3f3d187f68bdd0d8ff7056ba456f5f58/backend/src/CDMP/App/Error.hs#L53-L54